### ISSUE TITLE
Add fast wavenet generation.

### ISSUE THREAD

**Issue by tomlepaine** [ibab/tensorflow-wavenet#26]
@ibab I noticed you saw the efficient wavenet generation implementation I wrote with my friends:

https://github.com/tomlepaine/fast-wavenet

Can we help you add it to `tensorflow-wavenet`?

**Comment by ibab** [comment]
I was actually about to write you a mail :)
I'd be very happy to accept a PR if someone is interested in porting it over to this repo's implementation (Shouldn't be too difficult, as the core structure of the network is quite similar).
I should also be able to do it myself if necessary.

I noticed that your repo is licensed under GPL 3, which would be in conflict with this repo's MIT license if we were to copy code, though.

In the future I'd like to turn this repo into more of a community effort and move it to a Github organization, or the TensorFlow `models` repo.
Do you have any ideas or preferences as to where it should end up?

**Comment by tomlepaine** [comment]
Great! I can start working on it tomorrow, unless you want to do it.

And I'm flexible on the license. I'll change it to MIT.

The TensorFlow `models` repo sounds good to me.

**Comment by ibab** [comment]
Awesome! Feel free to comment here if you have a question or if there's something I can do.

**Comment by tomlepaine** [comment]
Update: Just about finished with the port. Will clean up the code and run some speed tests tonight, then make a pull request.

**Comment by tomlepaine** [comment]
Sorry for the delay!

It seems like the way we implemented our queue had a bug that made the states we pop incorrect a small percent of the time. We aren't sure of the exactly cause, but we think it is related to `tf.scatter_update`. We switched from our queue to `tf.FIFOQueue` and it seemed to fix the bug (and slightly improved our speed :tada:). Now (based on our tests) the fast generation exactly reproduces outputs and states of the slow generation.

We will make a pull request after lunch.
